drivers/flash/flash_rpc/flash_rpc_controller.c

```diff
@@ -0,0 +1,240 @@
+		LOG_ERR("Unable to decode result");
+		*result = -EINVAL;
+	}
+}
+
+static bool encode_flash_msg(struct nrf_rpc_cbor_ctx *ctx, off_t *offset, void *ptr, size_t *len)
+{
+	NRF_RPC_CBOR_ALLOC(&flash_rpc_api, *ctx, CBOR_BUF_FLASH_MSG_SIZE);
+
+	if (!zcbor_uint32_put(ctx->zs, (uint32_t)*offset)) {
+		return false;
+	}
+
+	if (!zcbor_uint32_put(ctx->zs, (uint32_t)ptr)) {
+		return false;
+	}
+
+	return zcbor_uint32_put(ctx->zs, (uint32_t)*len);
+}
+
+#ifndef CONFIG_FLASH_RPC_SYS_INIT
+static void err_handler(const struct nrf_rpc_err_report *report)
+{
+	LOG_ERR("nRF RPC error %d ocurred. See nRF RPC logs for more details.",
+	       report->code);
+	k_oops();
+}
+#endif
+
+int flash_rpc_init(const struct device *dev)
+{
+	int err;
+	int result;
+	struct nrf_rpc_cbor_ctx ctx;
+	const struct rpc_flash_config *dev_config = dev->config;
+
+	ARG_UNUSED(dev_config);
+
+#ifndef CONFIG_FLASH_RPC_SYS_INIT
+	err = nrf_rpc_init(err_handler);
+	if (err) {
+		LOG_ERR("Initializing nRF RPC failed: %d", err);
+		return -EINVAL;
+	}
+#endif
+
+	NRF_RPC_CBOR_ALLOC(&flash_rpc_api, ctx, CBOR_BUF_FLASH_MSG_SIZE);
+
+	err = nrf_rpc_cbor_cmd(&flash_rpc_api, RPC_COMMAND_FLASH_INIT, &ctx,
+			       rsp_error_code_handle, &result);
+	if (err != 0) {
+		LOG_ERR("Could not send RPC command: %d", err);
+		return err;
+	}
+
+	return result;
+}
+
+int flash_rpc_read(const struct device *dev, off_t offset, void *buffer, size_t len)
+{
+	ARG_UNUSED(dev);
+	int err;
+	int result;
+
+	struct nrf_rpc_cbor_ctx ctx;
+
+	if (buffer == NULL || len == 0) {
+		return -EINVAL;
+	}
+
+	if (!encode_flash_msg(&ctx, &offset, buffer, &len)) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("offset: %x, buffer_ptr: %p, len: %d", (uint32_t)offset, buffer, (uint32_t)len);
+
+	err = nrf_rpc_cbor_cmd(&flash_rpc_api, RPC_COMMAND_FLASH_READ, &ctx,
+				flash_rpc_get_rsp, &result);
+	if (err) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("Read command sent: %d", result);
+
+	return result;
+}
+
+int flash_rpc_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+	int err;
+	int result;
+
+	struct nrf_rpc_cbor_ctx ctx;
+
+	if (data == NULL || len == 0) {
+		return -EINVAL;
+	}
+
+	if (!encode_flash_msg(&ctx, &offset, (void *)data, &len)) {
+		return -EINVAL;
+	}
+
+	err = nrf_rpc_cbor_cmd(&flash_rpc_api, RPC_COMMAND_FLASH_WRITE, &ctx,
+			flash_rpc_get_rsp, &result);
+
+	LOG_DBG("offset: %x, data_ptr: %p, len: %d", (uint32_t)offset, data, (uint32_t)len);
+	if (err) {
+		LOG_ERR("Failed to send command: %d", err);
+		return -EINVAL;
+	}
+
+	LOG_DBG("Write command sent: %d", result);
+
+	return result;
+}
+
+int flash_rpc_erase(const struct device *dev, off_t offset, size_t size)
+{
+	ARG_UNUSED(dev);
+	int err;
+	int result;
+
+	struct nrf_rpc_cbor_ctx ctx;
+
+	if (!encode_flash_msg(&ctx, &offset, NULL, &size)) {
+		return -EINVAL;
+	}
+
+	err = nrf_rpc_cbor_cmd(&flash_rpc_api, RPC_COMMAND_FLASH_ERASE, &ctx,
+				flash_rpc_get_rsp, &result);
+	if (err) {
+		return -EINVAL;
+	}
```

Comment:
I think that EIO should be returned here and LOG_DBG/LOG_ERR with message and the real value of err.